File: src/typescript/3.longest-substring-without-repeating-characters.ts
`Given a string s, find the length of the longest substring without repeating characters.`;

function lengthOfLongestSubstring(s: string): number {
  let letters = new Set<string>();
  let left = 0,
    right = 0,
    longest = 0;
  while (right < s.length) {
    if (!letters.has(s[right])) {
      letters.add(s[right]);
      longest = Math.max(longest, letters.size);
      right++;
    } else {
      letters.delete(s[left]);
      left++;
    }
  }
  return longest;
}

File: src/python/3.longest-substring-without-repeating-characters.py
# Given a string s, find the length of the longest substring without repeating characters.


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # Time Complexity: O(n)
        # Space Complexity: O(n)
        if not s:
            return 0
        left, right = 0, 0
        max_len = 0
        seen = set()
        while right < len(s):
            while s[right] in seen:
                seen.remove(s[left])
                left += 1
            seen.add(s[right])
            max_len = max(max_len, right - left + 1)
            right += 1
        return max_len
